libraries/stdlib/osl/stdosl.h

```diff
@@ -673,10 +715,18 @@ closure color thin_film_bsdf(float thickness, float ior) BUILTIN;
 ​
 // Constructs an EDF emitting light uniformly in all directions.
 //
+//  \param  emittance   Radiant emittance of light leaving the surface.
+//
 closure color uniform_edf(color emittance) BUILTIN;
 
 // Constructs an EDF emitting light inside a cone around the normal direction.
 //
+//  \param  emittance   Radiant emittance of light leaving the surface.
+//  \param  N           Cone direction vector.
```

Comment:
Just curious what the role of `N` is here. Bump mapped emitters? I haven't seen any mesh light implementation that tries to accurately model the role of shading normals - let alone bump-mapped shading normals. I'm curious if anyone else feels like this is worth requiring.

For example - this introduces all the inconsistencies of shading normals to the emissive context. Some portion of the hemisphere around `N` will be self-shadowed against the hemisphere around `Ng`.

We had closures like this in the original OSL spec a long time ago but wound up removing them because they were hard to implement exactly. Besides, controls like spot angles are usually desired relative to the center of the light -- not across its area. The angular falloff of a spotlight typically should not change when you change the area of the light. But when described as a closure there would be an implicit "blur" by the surface area of the mesh you've assigned to.

Reply:
+1

Reply:
N here is mainly to set the direction of the cone when used in a light shader context when the shader is attached to a locator and not a surface. However, since the locator needs a transform anyway this is a bit redundant as we could instead define a single fixed direction and just add rotation to the transform.

In addition, for the same reasons we removed the measured_edf closure from this draft we could remove this one as well. It's main reason for being part of the Mtlx library is for defining light shaders for spot lights. As with the measured_edf closure this is supported in Nvidia MDL. But since OSL is currently not designed to describe light shaders with closures this way we can just remove this for now, and take a new look at this later if really needed.

Reply:
Thanks for those additional details! Makes sense, and I have removed the conical_edf as well for now.

There is the whole subject of lights standardization that I suppose we will get to eventually, including what's done in USD / UsdLux etc. So let's defer this to a later time.

Reply:
Nerdery: Technically, IES profiles describe what happens at a distance from the light that is 5x the lights own largest physical size axis. Meaning, they do not define near-field behavior, only far-field behavior.
I've been through this exact mess in about 3 renderers.... 

Some cheat one way, ny just applying the IES profile per sample, and hope the emitting area is small enough so that the distant result is "okay".
Some cheat the other way, and somehow apply distant magic like @fpsunflower mentions. 

Neither solution is ideal, THB. And out in reality, the emission profile of the light is a complicated result from it's reflector geometry and internal shadowing surfaces, and ain't nobody wanna model *that* for real :P